Implement resizable SABs for Wasm memories

ast-hugger opened the pull request:
#### e65b5874938c034f2d03b029d67afb64dff45eca
<pre>
Implement resizable SABs for Wasm memories
<a href="https://bugs.webkit.org/show_bug.cgi?id=297071">https://bugs.webkit.org/show_bug.cgi?id=297071</a>
<a href="https://rdar.apple.com/147437929">rdar://147437929</a>

Reviewed by Yusuke Suzuki.

This patch adds the implementation of the .toFixedLengthBuffer()
and .toResizableBuffer() methods of WebAssembly.Memory.prototype
as specified by <a href="https://webassembly.github.io/threads/js-api/index.html#memories.">https://webassembly.github.io/threads/js-api/index.html#memories.</a>

Availability of this feature is controlled by the option `useWasmMemoryToBufferAPIs`.
It is set to false on landing, except for the tests specifically targeting this,
such as `imported/w3c/web-platform-tests/wasm/jsapi/idlharness.any`.

Notable changes:

  - ArrayBuffer is given the ability to point at the Wasm memory it represents.
    This is used when the buffer is resizable non-shared, so that growth requests
    initiated by the buffer are routed to the memory. (In the shared case that
    is not needed because growing is always done by SharedArrayBufferContents).

  - To support the above, Wasm::Memory is made CanMakeWeakPtr.

  - JSWebAssemblyMemory::growSuccessCallback sends a refresh notification to
    its associated buffer if it&apos;s resizable non-shared, to update its length.
    Again, this explicit update is needed because in that setup there is no
    SharedArrayBufferContents common to the buffer and the memory.

  - The ArrayBuffer::isFixedLength() function is added, not because it&apos;s strictly
    necessary, but to match the IsFixedLengthArrayBuffer abstract operations in the spec,
    making it harder to make a mistake.

The expectations of WPT jsapi/memory tests are updated to expect the tests to pass.

Canonical link: <a href="https://commits.webkit.org/298955@main">https://commits.webkit.org/298955@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3147d0243b4cbcdee46a02eea73a25cbc2906c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36936 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27565 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123364 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69251 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7d3e1df4-d7ca-41bd-a771-b03ba77fa1a8) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37632 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45523 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88996 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43665 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ad46a3c8-34df-401b-ad3b-f9254c1a5915) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120199 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29948 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105156 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69504 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29008 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23270 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67038 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109371 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23452 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126486 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115773 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44163 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97667 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44519 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101392 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97462 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24820 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42806 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20748 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40513 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44036 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49695 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144473 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43492 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37178 "Failed to checkout and rebase branch from PR 49086") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46837 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45188 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -431,17 +451,21 @@ void ArrayBuffer::notifyDetaching(VM& vm)
     auto shared = m_contents.m_shared;
     if (!shared) [[unlikely]]
         return makeUnexpected(GrowFailReason::GrowSharedUnavailable);
-    auto result = shared->grow(vm, newByteLength);
```
Let's make it a bit more readable by using,
```
bool requirePageMultiple = isWasmMemory();
auto result = shared->grow(vm, newByteLength, requirePageMultiple);
```

Constellation commented on Source/JavaScriptCore/runtime/ArrayBuffer.h:
```diff
@@ -359,6 +371,7 @@ class ArrayBuffer final : public GCIncomingRefCounted<ArrayBuffer> {
 private:
     Checked<unsigned> m_pinCount { 0 };
     bool m_isWasmMemory { false };
```
Maybe you need to move this information to ArrayBufferContents (with a bit careful manner)? Because we support sharing SharedArrayBuffer between multiple workers, and SerializedScriptValue copied ArrayBufferContents to the other workers, and I guess that we need to keep this "originated from wasm memory" information is necessary.

ast-hugger replied:
This field is only really used when the buffer is resizable non-shared. The only use is in `ArrayBuffer::resize`. We need it because unlike with shared buffers/memories where we can just tell the `SharedArrayBufferContents` to grow and have the change visible in both the buffer and the memory, a non-shared buffer doesn't have a `SABC`. So instead a resizable non-shared buffer knows its associated Wasm memory, and to resize it asks the memory to grow and then the buffer is updated to match by `growSuccessCallback`/`refreshAfterWasmMemoryGrow`.

Maybe I can make this more apparent by only calling `setAssociatedWasmMemory` for resizable non-shared buffers, and having an assert that the buffer is of that kind.

With these constraints, I think not copying that information when serializing is the right thing to do since it only matters for non-shared buffers.

Constellation commented on Source/JavaScriptCore/wasm/js/WebAssemblyMemoryPrototype.cpp:
```diff
@@ -121,11 +144,21 @@ Structure* WebAssemblyMemoryPrototype::createStructure(VM& vm, JSGlobalObject* g
+        Identifier toResizableBufferName = Identifier::fromString(vm, "toResizableBuffer"_s);
+        JSFunction* toResizableBufferFunction = JSFunction::create(vm, globalObject, 0, toResizableBufferName.string(), webAssemblyMemoryProtoFuncToResizableBuffer, ImplementationVisibility::Public);
+        putDirectWithoutTransition(vm, toResizableBufferName, toResizableBufferFunction, static_cast<unsigned>(PropertyAttribute::None));
+    }
```
Let's use `JSC_NATIVE_FUNCTION_WITHOUT_TRANSITION`.

Constellation reviewed (approved):
r=me with comment.

ast-hugger committed:
Implement resizable SABs for Wasm memories
https://bugs.webkit.org/show_bug.cgi?id=297071
rdar://147437929

Reviewed by Yusuke Suzuki.

This patch adds the implementation of the .toFixedLengthBuffer()
and .toResizableBuffer() methods of WebAssembly.Memory.prototype
as specified by https://webassembly.github.io/threads/js-api/index.html#memories.

Availability of this feature is controlled by the option `useWasmMemoryToBufferAPIs`.
It is set to false on landing, except for the tests specifically targeting this,
such as `imported/w3c/web-platform-tests/wasm/jsapi/idlharness.any`.

Notable changes:

  - ArrayBuffer is given the ability to point at the Wasm memory it represents.
    This is used when the buffer is resizable non-shared, so that growth requests
    initiated by the buffer are routed to the memory. (In the shared case that
    is not needed because growing is always done by SharedArrayBufferContents).

  - To support the above, Wasm::Memory is made CanMakeWeakPtr.

  - JSWebAssemblyMemory::growSuccessCallback sends a refresh notification to
    its associated buffer if it's resizable non-shared, to update its length.
    Again, this explicit update is needed because in that setup there is no
    SharedArrayBufferContents common to the buffer and the memory.

  - The ArrayBuffer::isFixedLength() function is added, not because it's strictly
    necessary, but to match the IsFixedLengthArrayBuffer abstract operations in the spec,
    making it harder to make a mistake.

The expectations of WPT jsapi/memory tests are updated to expect the tests to pass.

Canonical link: https://commits.webkit.org/298955@main
